### DISCUSSION TITLE
Assign dynamic property-name in context for referencing actors

### DISCUSSION THREAD

**Discussion by Wastl77** [statelyai/xstate#3965] [Q&A] [answered]
Hi,

is there a way to assign the property name dynamically in the context for referencing actors?

This code works: 

```
on: {
          createResource: {
            actions: [
              assign({
                testref: (_, event) =>
                  spawn(resourceMachine, event.payload.callsign),
              })
            ],
          },
}
```

But this code works not:

```
on: {
          createResource: {
            actions: [
              (_, event) => {
                assign({
                  [event.payload.callsign]: (_, event) =>
                    spawn(resourceMachine, event.payload.callsign),
                })
              }
            ],
          },
}
```

Any ideas to set the property name dynamically in the context?

Thank you!

Greetings Sebastian

**Comment by davidkpiano** [comment, marked as answer]
The `assign(...)` action creator only returns an assign object which the machine uses to actually do the assignment. You can do this instead:

```ts
actions: assign((_, event) => {
  return {
    [event.payload.callsign]: spawn(...)
  }
})
```

**Reply by Wastl77** [reply]
Thank you for the quick reply! It works!

I have another question though, is there a way to invoke a promise, which results in an array of objects, and then in the onDone event iterate over the array and run the action mentioned above for each object in the array? (So that event.payload.callsign is then event.data.callsign)

Thank you!

**Reply by alexamy** [reply]
@Wastl77 Construct assign object from promise result array, like so:
```typescript
// event.data is [{ id: '', name: '' }, ...]
actions: assign((_, event) => {
  const actors = event.data.map(obj => [obj.id, obj.name]);
  return Object.fromEntries(actors);
}),
```
More complete example [here](https://codesandbox.io/s/action-on-array-from-promise-urvrkg?file=/index.js).

**Reply by Wastl77** [reply]
Hello Alexander,

thanks for your suggestion, i'm currently in Hospital and try out as soon as possible. I will give you feedback then.

Thanks!

**Reply by Wastl77** [reply]
@alexamy 👍  It works exactly as i need it! Thank you very much for the tip!
